controlplane/kubeadm/internal/filters.go

```diff
@@ -189,11 +189,16 @@ func matchClusterConfiguration(kcp *controlplanev1.KubeadmControlPlane, machine
 	if machineClusterConfig == nil {
 		machineClusterConfig = &bootstrapv1.ClusterConfiguration{}
 	}
-	kcpLocalClusterConfiguration := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration
+
+	kcpLocalClusterConfiguration := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DeepCopy()
```

Comment:
nit: to avoid the deep copy you could just
```go
machineClusterConfig.DNS = kcpLocalClusterConfiguration.DNS

Reply:
Thanks for your suggestion, it has been fixed.